lib/aiken/math.ak

```diff
@@ -155,6 +157,47 @@ test pow_513_3() {
   pow(513, 3) == 135005697
 }
 
+test pow_2_4() {
+  pow(2, 4) == 16
+}
+
+test pow_2_42() {
+  pow(2, 42) == 4398046511104
+}
+
+/// Calculates the power of 2 for a given exponent `e`. Much cheaper than
+/// using `pow(2, _)`.
+///
+/// ```aiken
+/// math.pow2(0) == 1
+/// math.pow2(1) == 2
+/// math.pow2(4) == 16
+/// math.pow2(42) == 4398046511104
+/// ```
+pub fn pow2(e: Int) -> Int {
+  if e <= 7 {
+    builtin.index_bytearray(#[1, 2, 4, 8, 16, 32, 64, 128], e)
+  } else {
+    256 * pow2(e - 8)
```

Comment:
Random question, is this tail called optimized by the compiler? It's not per-se tail call, but this kind of expression can still be TCO'd in pure language with TRMC (cf https://jfmengels.net/modulo-cons/)

Reply:
It also seem like the function is not dealing with negative integers? and does not have exponential cost decrease with very big exponents (since there is no branch for square computations).
I'd imagine something like 
```
if e <= 0 { ... }
else if e <= 7 { ... }
else if e <= 63 { 256 * pow2(e - 8) }
else { the square stuff }
```

Reply:
The negative case is handled "implicitly" at the moment: it fails. As `index_bytearray` will crash if the index is out of bound (so `e` negative). The behavior is likely fine since we don't have floating numbers in the VM any negative exponent would yield `0`. We could also just return `0` actually instead of failing. That's one extra comparison on each 'loop' but that sounds fairly okay. 

The square case is a good idea too. Will do 👍

Reply:
You could do 1 initial check and then have a loop function.

Reply:
Ok, failing is rather ok I guess, but I asked because currently it is not consistent with the other `pow` function. Regarding the square stuff, I'm wondering what are the memory/compute trade offs and the price of the square branch, since it has `% 2` and `/ 2` computations. I set the limit at 64 but I guess it could be something else. Also wondering what would be the perf&memory profile of something like:
```
if e < 16 { the index stuff }
else { the square stuff }
```
without at all the `256 * pow2(e-8)` branch.
Anyway, I'm probably overthinking this. Just thought it would be super cool to produce like graphs of the cost for `e` between 0 and like 1000 for each impl.

Reply:
let's keep this simple for the stdlib

Reply:
So, after playing a bit with it, turns out that the squared exponentiation isn't easy to combine with the byte-array trick as we need to square the base values. I tried various things but then the function itself gets too complex and we loose the benefits of the optimization. 

So, I've added one more increment to that for exponents up to 256 the function is still faster than `pow`; and for any larger exponent, `pow` is already doing proper squaring method and gets actually faster.